### PR TITLE
mockgcp: support for KMSCryptoKey

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

File: mockgcp/apply-proto-patches.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+set -x
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
+cd ${REPO_ROOT}/mockgcp
+
+cd tools/patch-proto
+
+# Use our proto patch tool to add the missing maintenanceUpdatePolicy field for alloydb
+go run . --file ${REPO_ROOT}/mockgcp/third_party/googleapis/google/cloud/alloydb/v1beta/resources.proto --message Cluster <<EOF
```
Comment (yuwenma) [inline mockgcp/apply-proto-patches.sh:30]: Mind sharing more context about this?

File: pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml [hunk 2]
```diff
@@ -41,14 +41,13 @@ spec:
   clusterType: PRIMARY
   continuousBackupConfig:
     enabled: false
-    recoveryWindowDays: 14
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml:44]: Do you know why this is changed?

File: pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml [hunk 3]
```diff
@@ -62,8 +61,6 @@ spec:
   networkConfig:
     networkRef:
       name: computenetwork-${uniqueId}
-  networkRef:
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml:65]: Is this a dup field with the networkConfig.networkRef? I'm wondering why the golden object is different (I was expecting the presubmit check to block the original change).

**Review by yuwenma — COMMENTED** [review]
This PR seems to contain several different feats and chores (may come from different PRs). Do you mind sharing the PR review?

**Comment by justinsb** [comment]
> This PR seems to contain several different feats and chores (may come from different PRs). Do you mind sharing the PR review?

I don't understand "Do you mind sharing the PR review?" - can you clarify?

But yes, when we enabled KMSCryptoKey in mockgcp, that meant we now started running a lot more tests against mockgcp - not just KMSCryptoKey tests - because they were "missing" this one resource.  There are a bunch of fixes to mockgcp to enable those additional tests to pass.

The challenge is that because those tests don't start running until we have KMSCryptoKey, I can't easily split them out (I can send them as separate PRs, but we won't get any test coverage without KMSCryptoKey)

File: mockgcp/apply-proto-patches.sh (hunk 1 above)
Comment (justinsb) [reply mockgcp/apply-proto-patches.sh:30]: Yes - good call.  Our public protos are missing the maintenanceUpdatePolicy field (I filed an internal bug for that).

The tool can patch the Protos.  That tool is sort of overkill (it uses treesitter), but it is something we're doing a lot of anyway because of CSV exporting for AI and when we're generating our types/mappers/etc.

What this lets us do is make progress when the public protos are not complete.  I think that's better than trying to maintain patches etc.  If we had more GitHub repos, we would probably just carry a fork of the public googleapis protos, but creating a repo is annoyingly difficult for us!

I'll add more comments etc.

File: pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml (hunk 2 above)
Comment (justinsb) [reply pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml:44]: I think it's because we weren't running this test previously (because it depends on KMSCryptoKey).

It's a gap in the mockgcp for AlloydbCluster.  It should default recoveryWindowDays to 14 I think.

The tricky thing is ... do we want to do this in this PR?  If we do, the PR becomes even bigger.  I do think once we get support for KMSCryptoKey in, we will want to do a `record-gcp` followed by a `compare-mock` on the AlloyDB fixtures, to fix up any loose ends in mockgcp here.

File: pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml (hunk 3 above)
Comment (justinsb) [reply pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml:65]: Good question, I don't know.  Maybe @anhdle-sso knows (is this what you meant by sharing the PR review?)

The presubmits weren't running here before, because we didn't have a mock for KMSCryptoKey

File: pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml (hunk 2 above)
Comment (yuwenma) [reply pkg/test/resourcefixture/testdata/basic/alloydb/v1beta1/alloydbcluster/fullalloydbcluster/_generated_object_fullalloydbcluster.golden.yaml:44]: I see. That makes sense.

Agree to complete the AlloyDb later once the KMSCryptoKey is in.

**Comment by yuwenma** [comment]
> > This PR seems to contain several different feats and chores (may come from different PRs). Do you mind sharing the PR review?
> 
> I don't understand "Do you mind sharing the PR review?" - can you clarify?
> 
> But yes, when we enabled KMSCryptoKey in mockgcp, that meant we now started running a lot more tests against mockgcp - not just KMSCryptoKey tests - because they were "missing" this one resource. There are a bunch of fixes to mockgcp to enable those additional tests to pass.
> 
> The challenge is that because those tests don't start running until we have KMSCryptoKey, I can't easily split them out (I can send them as separate PRs, but we won't get any test coverage without KMSCryptoKey)

My bad (dumb brain after a working week). I was trying to ask to split this PR into multiple individual ones that each can focus on one MockGCP change, so the code can go in faster and won't have much merge conflicts. Mostly because I have several questions on different resources.

File: mockgcp/Makefile [hunk 4]
```diff
@@ -31,8 +18,13 @@ gen-proto: tools
 	git clone https://github.com/googleapis/googleapis.git third_party/googleapis || (cd third_party/googleapis && git reset --hard origin/master && git pull)
 	mkdir -p ./generated
 
+	./apply-proto-patches.sh
 	./fixup-third-party.sh
 
+	make gen-proto-no-fixup
+
+.PHONY: gen-proto-no-fixup
+gen-proto-no-fixup:
```
Comment (yuwenma) [inline mockgcp/Makefile:27]: I like this change! Shall this make rule contain the `mkdir -p ./generated`.

File: mockgcp/Makefile [hunk 5]
```diff
@@ -31,8 +18,13 @@ gen-proto: tools
 	git clone https://github.com/googleapis/googleapis.git third_party/googleapis || (cd third_party/googleapis && git reset --hard origin/master && git pull)
 	mkdir -p ./generated
 
+	./apply-proto-patches.sh
```
Comment (yuwenma) [inline mockgcp/Makefile:21]: If we want to make the apply-proto-patches a permanent solution (I think we do), then maybe call out that this is an one-off run since the patch always appends the proto?

File: mockgcp/mockcontainer/cluster.go [hunk 6]
```diff
@@ -217,6 +217,43 @@ func (s *ClusterManagerV1) UpdateCluster(ctx context.Context, req *pb.UpdateClus
 	})
 }
 
+func (s *ClusterManagerV1) SetLabels(ctx context.Context, req *pb.SetLabelsRequest) (*pb.Operation, error) {
```
Comment (yuwenma) [inline mockgcp/mockcontainer/cluster.go:220]: Also want to ask for more context about this change since it seems to be unrelated to KMS and alloydb (maybe I"m wrong). If this is an improvement to container cluster only, can we move it to a separate PR? (I thought I reviewed this code somewhere, so it would be really nice to have it go)

File: mockgcp/mockkms/cryptokeyversion.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+func (r *kmsServer) CreateCryptoKeyVersion(ctx context.Context, req *pb.CreateCryptoKeyVersionRequest) (*pb.CryptoKeyVersion, error) {
+	parentName, err := r.parseCryptoKeyName(req.GetParent())
+	if err != nil {
+		return nil, err
+	}
+
+	id := strconv.FormatInt(time.Now().UnixNano(), 10)
+
+	// The server-generated crypto key version name is the concatenation of the parent crypto key's
```
Comment (yuwenma) [inline mockgcp/mockkms/cryptokeyversion.go:92]: Thank you for adding the comment!

File: mockgcp/mocksql/sqlinstance.go [hunk 8]
```diff
@@ -81,6 +81,8 @@ func (s *sqlInstancesService) Insert(ctx context.Context, req *pb.SqlInstancesIn
 		obj.MaintenanceVersion = "SQLSERVER_2017_EXPRESS_CU31_GDR.R20231029.00_02"
 	case pb.SqlDatabaseVersion_POSTGRES_9_6:
 		obj.DatabaseInstalledVersion = "POSTGRES_9_6"
+	case pb.SqlDatabaseVersion_POSTGRES_15:
+		obj.DatabaseInstalledVersion = "POSTGRES_15"
 	default:
```
Comment (yuwenma) [inline mockgcp/mocksql/sqlinstance.go:86]: Some question: I guess there is a reason to add this change here.

File: mockgcp/tools/patch-proto/main.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	sitter "github.com/smacker/go-tree-sitter"
```
Comment (yuwenma) [inline mockgcp/tools/patch-proto/main.go:28]: Just curious, why do you say the treesitter is an overkill? At the beginning I thought it means a lot of code, but it seems to be fine. Does this run take a long time to complete?

File: mockgcp/apply-proto-patches.sh (hunk 1 above)
Comment (yuwenma) [reply mockgcp/apply-proto-patches.sh:31]: Thank you! Yeah, I saw the internal bug. May I ask how you find this missing field?

**Comment by yuwenma** [comment]
/lgtm
/approve

Have a few questions want to follow up, no blockers
